pkg/zng/docs/spec.md

```diff
@@ -1,96 +1,268 @@
+can be mapped onto ZNG and recovered by decoding
+that ZNG back into JSON.
 
-ZSON is a UTF-8 encoded stream of "lines" where each line is terminated by a
-newline.  Each line is either a directive or a value.
+> TBD: ZNG is also interchangeable with Avro...
 
-Directives and values, in turn, come in two flavors: regular and legacy.
-Thus, there are four types of lines:
-* regular directives,
-* regular values,
-* legacy directives, and
-* legacy values.
+The ZNG design [is also motivated by](./rationale.md)
+and maintains compatibility with the original
+[default Zeek log format](https://docs.zeek.org/en/stable/examples/logs/).
 
-Any line that begins with character `#` is a directive.
-All other lines are values.  If a value line begins with `#`, then this
-character must be escaped.
-(Such lines can only exist as legacy values and not regular values
-as regular values always begin with an integer descriptor.)
+## 1. The ZNG data model
 
-Directives indicate how subsequent values in the ZSON stream are interpreted.
-A value is a regular value if the most recent preceding directive in the stream
-is a regular directive; otherwise, it is a legacy value.
+ZNG encodes a sequence of one or more typed data values comprising a stream.
+The stream of values is interleaved with control messages
+that provide type definitions and other metadata.  The type of
+a particular data value is specified by its "type code", which is an
+integer identifier representing either a built-in type or
+a dynamic type definition that occurred previously in the stream.
 
-Any line beginning with `#` that does not conform with the syntax of a
-directive is an error.
-When errors are encountered parsing ZSON, an implementation should return a
-corresponding error and allow ZSON parsing to proceed if desired.
+The ZNG type system comprises the standard set of scalar types like integers,
+floating point, strings, byte arrays, etc. as well as composite types
+like records, arrays, and sets arranged from other types.
+
+For example, a ZNG stream representing the single string "hello world"
+looks like this:
+```
+9:hello, world
+```
+Here, the type code is the integer "9" representing the string type and the data
+value "hello, world" is an instance of string.
+
+ZNG gets more interesting when data types are interleaved in the stream.
+For example,
+```
+9:hello, world
+4:42
+9:there's a fly in my soup!
+9:no, there isn't.
+4:3
+```
+where type code 4 represents an integer.  This encoding represents the string of
+values:
+```
+"hello, world"
+42
+"there's a fly in my soup!"
+"no, there isn't."
+3
+```
+ZNG streams are often comprised as a sequence of records, which works well to provide
+an efficient representation of structured logs.  In this case, a new type code is
+needed to define the schema for each distinct record.  To define a new
+type, the "#" syntax is used.  For example,
+logs from the open-source Zeek system might look like this
+```
+#23:ip=addr
```

Comment:
this `ip=addr` type definition on the universal zng constant isn't used in this example

Reply:
Correct.  Went back into WIP mode....

Reply:
Will address